src/react/controllers/map-controller.js

```diff
@@ -109,18 +113,122 @@ export default class MapController extends PureComponent {
+    const {isDragging = false} = interactiveState;
     if (isDragging !== this.state.isDragging) {
       this.setState({isDragging});
     }
   }
 
+  _extractViewportFromProps(props) {
+    return {
+      width: props.width,
+      height: props.height,
+      latitude: props.latitude,
+      longitude: props.longitude,
+      zoom: props.zoom,
+      bearing: props.bearing,
+      pitch: props.pitch,
+      altitude: props.altitude
+    };
+  }
+
+  _animateViewportProp(nextProps) {
+    if (this.props.animateViewport !== true) {
+      return;
+    }
+    const startViewport = this._extractViewportFromProps(this.props);
+    const endViewport = this._extractViewportFromProps(nextProps);
+    if (this._viewportChanged(startViewport, endViewport)) {
+      if (this.state.animationInterval) {
+        clearInterval(this.state.animationInterval);
+      }
+      const animationInterval = setInterval(() => this._updateViewport(), 50);
+      this.setState({
+        animationInProgress: true,
+        animationT: 0.0,
+        animationStartViewport: startViewport,
+        animationEndViewport: endViewport,
+        animationInterval,
+        animatedViewport: startViewport
+      });
+    }
+  }
+
+  _viewportChanged(startViewport, endViewport) {
+    for (const p of VIEWPORT_ANIMATE_PROPS) {
+      if (startViewport[p] !== undefined &&
+        endViewport[p] !== undefined &&
+        startViewport[p] !== endViewport[p]) {
+          // console.log(`viewport changed : ${p}: ${startViewport[p]}->${endViewport[p]}`);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  _updateViewport() {
+    // console.log(`Animating viewport t: ${this.state.animationT}`);
+
+    const animatedViewport = Object.assign({}, this.state.animationEndViewport);
+    const t = this.state.animationT;
+    for (const p of VIEWPORT_ANIMATE_PROPS) {
+      const startValue = this.state.animationStartViewport[p];
+      const endValue = this.state.animationEndViewport[p];
+      animatedViewport[p] = this._interpolate(startValue, endValue, t);
+    }
+
+    if (t <= 1.0) {
+      this.setState(prevState => ({
+        animationT: prevState.animationT + 0.01,
+        animatedViewport
+      }));
+    } else {
+      this._endAnimation();
+    }
+  }
+
+  _interpolate(start, end, t) {
+    // TODO: this method can be a prop, like easingFunction
+    return t * end + (1 - t) * start;
+  }
+
+  _endAnimation() {
+    // console.log('End Animation');
+    clearInterval(this.state.animationInterval);
+    this.setState({
+      animationT: 0,
+      animationInterval: null,
+      animationInProgress: false,
+      animationStartState: null,
+      animationEndState: null,
+      animatedViewport: null
+    });
+  }
+
+  _recursiveCloneChildren(children, viewport) {
+    return Children.map(children, child => {
+      if (!isValidElement(child)) {
+        return child;
+      }
+      // TODO: we need to filter chidren and only update those that require
```

Comment:
One technique could be to look for viewport props on the children. If they don't have the props you are animating, no need to clone them...

Reply:
If we decided to group all viewport props into `viewportState`, this gets easier, we can just check if a child contains this prop, instead of checking 7/8 individual props.